File: apps/web/src/types/task.ts
/**
 * Task Type Definitions
 * Phase 2 Chunk 2 - Tasks Full Feature Set
 */

/**
 * Task lifecycle status
 */
export type TaskStatus = 'pending' | 'in_progress' | 'completed';

/**
 * Task priority levels
 */
export type TaskPriority = 'high' | 'medium' | 'low' | null;

/**
 * Task entity - represents a user's to-do item
 */
export interface Task {
  id: string;  // UUID
  user_id: string;  // UUID
  title: string;  // Required, max 500 chars
  description: string | null;  // Optional, max 5000 chars
  status: TaskStatus;
  priority: TaskPriority;
  tags: string[];  // Array of user-defined tags
  due_date: string | null;  // ISO 8601 timestamp or null
  completed: boolean;
  created_at: string;  // ISO 8601 timestamp
  updated_at: string;  // ISO 8601 timestamp
}

/**
 * Task creation request payload
 */
export interface TaskCreateRequest {
  title: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  tags?: string[];
  due_date?: string;  // ISO 8601 timestamp
}

/**
 * Task update request payload (partial update)
 */
export interface TaskUpdateRequest {
  title?: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  tags?: string[];
  due_date?: string;  // ISO 8601 timestamp
}

/**
 * Task list filters for querying tasks
 */
export interface TaskFilters {
  status?: TaskStatus;
  priority?: TaskPriority;
  tags?: string;  // Comma-separated string
  search?: string;
  sort?: TaskSortOption;
  page?: number;
  limit?: number;
}

/**
 * Sort options for task lists
 */
export type TaskSortOption =
  | 'created_desc'
  | 'created_asc'
  | 'due_date_asc'
  | 'due_date_desc'
  | 'priority_asc'
  | 'priority_desc';

/**
 * Paginated task list response
 */
export interface TaskListResponse {
  tasks: Task[];
  total: number;
  page: number;
  limit: number;
}

/**
 * Task form validation errors
 */
export interface TaskValidationErrors {
  title?: string;
  description?: string;
  priority?: string;
  tags?: string;
  due_date?: string;
}

/**
 * Validate task title
 */
export function validateTitle(title: string): string | null {
  if (!title || !title.trim()) {
    return 'Title cannot be empty';
  }
  if (title.length > 500) {
    return 'Title must be 500 characters or less';
  }
  return null;
}

/**
 * Validate task description
 */
export function validateDescription(description: string): string | null {
  if (description && description.length > 5000) {
    return 'Description must be 5000 characters or less';
  }
  return null;
}

/**
 * Validate task priority
 */
export function validatePriority(priority: string | null): string | null {
  if (priority && !['high', 'medium', 'low'].includes(priority)) {
    return 'Priority must be high, medium, or low';
  }
  return null;
}

/**
 * Validate entire task form
 */
export function validateTaskForm(
  data: TaskCreateRequest | TaskUpdateRequest
): TaskValidationErrors {
  const errors: TaskValidationErrors = {};

  if ('title' in data && data.title !== undefined) {
    const titleError = validateTitle(data.title);
    if (titleError) errors.title = titleError;
  }

  if (data.description) {
    const descError = validateDescription(data.description);
    if (descError) errors.description = descError;
  }

  if (data.priority) {
    const priorityError = validatePriority(data.priority);
    if (priorityError) errors.priority = priorityError;
  }

  return errors;
}

File: apps/api/src/models/task.py
"""
Task Model
Phase 2 Chunk 2 - Tasks Full Feature Set
"""
from sqlmodel import SQLModel, Field, Column
from sqlalchemy import ARRAY, String, Text
from uuid import UUID, uuid4
from datetime import datetime, timezone
from typing import Optional
from pydantic import field_validator


class Task(SQLModel, table=True):
    """
    Task entity representing a user's to-do item.

    Attributes:
        id: Unique task identifier (UUID)
        user_id: Foreign key to users table
        title: Task title (required, max 500 chars)
        description: Optional detailed description (max 5000 chars)
        status: Task lifecycle status (pending, in_progress, completed)
        priority: Priority level (high, medium, low, or null)
        tags: Array of user-defined tags/categories
        due_date: Optional deadline timestamp (UTC)
        completed: Completion flag (synced with status via trigger)
        created_at: Task creation timestamp
        updated_at: Last modification timestamp (auto-updated via trigger)
    """

    __tablename__ = "tasks"

    id: UUID = Field(default_factory=uuid4, primary_key=True)
    user_id: UUID = Field(foreign_key="users.id", nullable=False, index=True)
    title: str = Field(sa_column=Column(Text, nullable=False))
    description: Optional[str] = Field(default=None, sa_column=Column(Text, nullable=True))
    status: str = Field(default="pending", nullable=False, max_length=20)
    priority: Optional[str] = Field(default=None, max_length=10)
    tags: list[str] = Field(default_factory=list, sa_column=Column(ARRAY(String), nullable=True))
    due_date: Optional[datetime] = Field(default=None)
    completed: bool = Field(default=False, nullable=False)
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), nullable=False
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc), nullable=False
    )

    @field_validator("title")
    @classmethod
    def title_not_empty(cls, v: str) -> str:
        if not v or not v.strip():
            raise ValueError("Title cannot be empty or only whitespace")
        v = v.strip()
        if len(v) > 500:
            raise ValueError("Title must be 500 characters or less")
        return v

    @field_validator("description", mode="before")
    @classmethod
    def description_max_length(cls, v):
        if v is not None and len(v) > 5000:
            raise ValueError("Description must be 5000 characters or less")
        return v

    @field_validator("tags", mode="before")
    @classmethod
    def trim_tags(cls, v):
        if isinstance(v, list):
            return [tag.strip() for tag in v if tag.strip()]
        return v

    class Config:
        """Pydantic configuration"""

        json_schema_extra = {
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "user_id": "650e8400-e29b-41d4-a716-446655440000",
                "title": "Write project proposal",
                "description": "Complete the Q1 project proposal document",
                "status": "pending",
                "priority": "high",
                "tags": ["work", "urgent", "client"],
                "due_date": "2026-01-10T00:00:00Z",
                "completed": False,
                "created_at": "2026-01-04T10:00:00Z",
                "updated_at": "2026-01-04T10:00:00Z",
            }
        }
